[pkg/util/log] Add zap wrapper over agent logger

mx-psi opened the pull request:
### What does this PR do?

Add [`zap`](https://pkg.go.dev/go.uber.org/zap) wrapper over the Agent's default logger.

### Motivation

Part of DataDog/architecture#892, to be able to log messages from OpenTelemetry Collector components.

### Additional Notes

We adapt some code from the zap library which is sadly not exported.

### Describe how to test your changes

Test that existing log code works properly (specially file, line and function should be correct)

### Checklist
<!-- Place an '[x]' (no spaces) in all applicable fields. Please remove unrelated fields. -->

- [x] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [x] The `need-change/operator` and `need-change/helm` labels has been applied if applicable.
- [x] The appropriate `team/..` label has been applied, if known.
- [x] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [x] The [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated if applicable.

Note: Adding GitHub labels is only possible for contributors with write access.

mx-psi committed:
[pkg/util/log] Add zap wrapper over agent logger

mx-psi committed:
[pkg/util/log] Add unit tests

mx-psi commented on pkg/util/log/zapcore_test.go:
```diff
@@ -0,0 +1,128 @@
+			desc:    "Debug (no fields, debug level)",
+			f:       func(l *zap.Logger) { l.Debug("Simple message") },
+			level:   "debug",
+			pattern: "\\[DEBUG\\] Write: \\(log/zapcore_test.go:\\d+\\) \\| Simple message",
```
I am not super happy with this way of writing the tests; it works but it's fragile: e.g. if you forget to escape a `|` the tests can easily pass on incorrect situations. I could just hardcode the line numbers which would be safer but more cumbersome to change.

mx-psi committed:
Minor fixes (fix test and avoid copy)

KSerrania commented on pkg/util/log/zapencoder.go:
```diff
@@ -0,0 +1,124 @@
+	return nil
+}
+
+func (e *encoder) OpenNamespace(ns string) {
+	if e.prefix == "" {
+		e.prefix = ns + "/"
+	} else {
+		e.prefix = e.prefix + ns + "/"
+	}
+}
```
Would the following be equivalent to the current implementation:
```suggestion
func (e *encoder) OpenNamespace(ns string) {
	e.prefix = e.prefix + ns + "/"
}
```

given that `e.prefix + ns == ns` if `e.prefix == ""`?

KSerrania commented on pkg/util/log/zapencoder_test.go:
```diff
@@ -0,0 +1,154 @@
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.desc, func(t *testing.T) {
+			enc := &encoder{}
+			tt.f(enc)
+			assert.Equal(t, tt.expected, enc.ctx, "Unexpected encoder output.")
+		})
+	}
```
Same here (renaming `tt` and `f`)

mx-psi replied:
Done

mx-psi committed:
Apply suggestions from code review

Co-authored-by: Kylian Serrania <kylian.serrania@datadoghq.com>

mx-psi committed:
[pkg/util/log] Remove unused test lines

These were left there by mistake

mx-psi committed:
[pkg/util/log] rename test variables

mx-psi commented on pkg/util/log/zapcore_test.go:
```diff
@@ -0,0 +1,131 @@
+			pattern := fmt.Sprintf("^%s$", tt.pattern)
+			assert.Regexp(t, pattern, strings.TrimSuffix(b.String(), "\n"))
+		})
+	}
```
Done!

mx-psi commented on pkg/util/log/zapcore_test.go:
```diff
@@ -0,0 +1,131 @@
+	logger.Info("with fields", zap.Int("val", 1))
+
+	loggerWithCtx := logger.With(zap.String("context", "this"))
+	loggerWithCtx.Warn("A warning", zap.Bool("val", false))
```
It's just that I forgot to remove them 😅, I deleted them now

KSerrania reviewed (approved):
LGTM

mx-psi committed:
[pkg/util/log] Add `NewZapCore`

We may want to use the core directly if we use the service.Collector struct.

mx-psi committed:
[pkg/util/log] Get correct file by setting stack depth

We make logs first-class by increasing the stack depth by 3
(e.g.: infocwithdepth + Write + Info) before each call.

When using this for external packages we may want to trim the paths
too, but this will come later.

mx-psi committed:
[pkg/util/log] `Sync` should flush

mx-psi committed:
Apply suggestions from code review

Co-authored-by: Kylian Serrania <kylian.serrania@datadoghq.com>

olivielpeau commented on pkg/util/log/zapcore.go:
```diff
@@ -0,0 +1,94 @@
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021-present Datadog, Inc.
+
+package log
```
I think `zapcore.go` and `zapencoder.go` would better fit in a new, separate package (maybe `pkg/util/log/zap`?), that would make for a cleaner interface with `pkg/util/log` with exported functions, and avoid adding zap as a dependency of `pkg/util/log`. WDYT?

mx-psi replied:
That makes sense to me, would we export the `tracecWithDepth`-like functions, then? (so, in your other comment, we would rename e.g. `tracecWithDepth` to `TracecStackDepth`?)

olivielpeau replied:
Yes, exactly

olivielpeau commented on pkg/util/log/log.go:
```diff
@@ -566,9 +570,13 @@ func Tracef(format string, params ...interface{}) {
 	logFormat(seelog.TraceLvl, func() { Tracef(format, params...) }, logger.tracef, format, params...)
 }
 
+func tracecWithDepth(message string, depth int, context ...interface{}) {
```
Although I prefer this name, for consistency with the existing functions (example `TraceStackDepth` I think we should call this `tracecStackDepth`), WDYT?

(ditto for all the other similar functions for the other log levels)

mx-psi committed:
Move to separate zap package
